### DISCUSSION TITLE
nginx proxy

### DISCUSSION THREAD

**Discussion by xscan** [pocketbase/pocketbase#105] [Q&A]
Configuration needs to be added
```
        proxy_set_header Connection '';
        proxy_http_version 1.1;
        chunked_transfer_encoding off;
        proxy_buffering off;
        proxy_cache off;
```

Examples
  ```

  location / {
          proxy_set_header Connection '';
          proxy_http_version 1.1;
          chunked_transfer_encoding off;
          proxy_buffering off;
          proxy_cache off;
          proxy_set_header  Host $http_host;
          proxy_set_header  X-Real-IP $remote_addr;
          proxy_set_header  X-Forwarded-For $proxy_add_x_forwarded_for;
          proxy_set_header  X-Forwarded-Proto http;
          proxy_set_header  X-Forwarded-Port $server_port;
          proxy_pass http://127.0.0.1:8090/; 
      }
```

**Comment by ganigeorgiev** [comment]
@xscan For the minimal setup this shouldn't be really necessary. Was there any particular reason why the above had to be added in your case? What nginx version are you using?

**Reply by xscan** [reply]
nginx 1.18.0 Failed to push without adding parameters sse

**Reply by ganigeorgiev** [reply]
Hm, I'll test it later again but the above shouldn't be required.

**Reply by ganigeorgiev** [reply]
@xscan I've just tested it and indeed the `proxy_http_version` directive is necessary for the SSE connections (it is even mentioned in the nginx docs http://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive).

I've updated the example in https://pocketbase.io/docs/going-to-production/#deploy-behind-reverse-proxy and added only the minimum recommended directives:
```
proxy_set_header Connection '';
proxy_http_version 1.1;
chunked_transfer_encoding off;
```

`proxy_cache` is disabled by default - http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache

`proxy_buffering` (and other buffer directives) should be changed only if the user really know what they are doing because disabling it may have side effect on the rate limiting and caching. It indeed could help for long lived connections but it's better to leave the decision to the user and not imply a default (see also https://www.nginx.com/blog/avoiding-top-10-nginx-configuration-mistakes/#proxy_buffering-off).

**Reply by xscan** [reply]
I can't access the backend after adding this configuration on my side
chunked_transfer_encoding off;

**Reply by ganigeorgiev** [reply]
Let me check again. I was testing with nginx in docker and maybe the mounted volume config was not loaded.

**Reply by xscan** [reply]
nginx retains these two configurations for the backend and sse to access properly

```
  proxy_set_header Connection '';
  proxy_http_version 1.1;
```

**Reply by xscan** [reply]
```
 location / {
        proxy_set_header Connection '';
        proxy_http_version 1.1;
        # chunked_transfer_encoding off;
        # proxy_buffering off;
        # proxy_cache off;
        proxy_set_header  Host $http_host;
        proxy_set_header  X-Real-IP $remote_addr;
        proxy_set_header  X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header  X-Forwarded-Proto http;
        proxy_set_header  X-Forwarded-Port $server_port;
        proxy_pass http://127.0.0.1:8090/; #此处改为 docker 服务的地址
    }
```

**Reply by ganigeorgiev** [reply]
@xscan Sorry for the delay. I've just tested it on nginx 1.18 and it works both with and without `chunked_transfer_encoding off;`. So I'm not sure what actually your issue is. Here is the config that I've tried:
```
server {
    listen 80;
    # server_name example.com;

    location / {
        # check http://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive
        proxy_set_header Connection '';
        proxy_http_version 1.1;
        chunked_transfer_encoding off; 

        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_pass http://127.0.0.1:8090;
    }
}
```

If configured, do you have any errors in the access logs?

**Reply by ganigeorgiev** [reply]
@xscan After some more tests, you may not need the chunked_transfer_encoding directive, but you may want to adjust the `proxy_read_timeout` directive since by default is 60s and if your client doesn't receive any SSE message during that time, the connection will be aborted (browsers has auto reconnect support for SSE). In the go code this timeout is 5 minutes but may change in future versions.

So, in other words, the below config should be a correct example of minimal nginx setup, supporting SSE (I'll update the docs later):
```nginx
server {
    listen 80;
    server_name example.com;

    location / {
        # check http://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive
        proxy_set_header Connection '';
        proxy_http_version 1.1;
        proxy_read_timeout 180s;

        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_pass http://localhost:8090;
    }
}
```

**Comment by MedLabs** [comment]
I'm trying to setup nginx to get pocketbase on https://mysite.com:8099
Tried a lot of combinations, and here is my nginx config file for pocketbase:

```
server {
server_name mysite.com
location / {
        proxy_set_header Connection '';
        proxy_http_version 1.1;
        proxy_read_timeout 180s;

        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_pass http://127.0.0.1:8090;
}
listen [::]:8099 ssl;
listen 8099 ssl;

ssl_certificate ......
ssl_certificate_key .....
......
}
```

I tried to run pocketdb instance on --http="127.0.0.1:8090" then on --https="127.0.0.1:8090"
changed proxy_pass to 127.0.0.1:8099 too but I couldn't find the wright config.

My mongodb backend is running well with a listen 5000 ssl (but there are index index.html on that config.

what is wrong with my config ? should I add root /path/to/pocketbase to my localtion ?

**Reply by ganigeorgiev** [reply]
@MedLabs If you are using nginx with your own SSL management, then you don't need the `--https` flag.

Starting the application with just `./pocketbase serve --http="127.0.0.1:8090"` should work just fine.

Make also sure that your configuration doesn't have any syntax errors. In the above example it seems that a `;` is missing at the end of the `server_name` directive. You may want to run `nginx -t` to ensure that everything is correct.

If everything with your config is OK, then you may try replacing `127.0.0.1:8090` with:
- `localhost:8090`
- `[::]:8090`
- directly with your public IP:8090
- bind to all interfaces using `0.0.0.0:8090`

This generally depends a lot on your server configurations and without any additional information or error logs, it is hard to tell what is wrong with the setup.

**Reply by MedLabs** [reply]
so if I remove the pocketbase nginx config file, I can just use --https option ?

**Reply by ganigeorgiev** [reply]
@MedLabs You need a reverse proxy like nginx if you want to run multiple applications on the same port. Usually on servers that run multiple public services you'll have nginx/apache/etc. that listens on 80/443 and proxies the requests based on your configurations to the corresponding applications via the `proxy_pass` directive.

In general, PocketBase doesn't need nginx and can be started with `--https` flag, which will generate a Let's encrypt certificate for you. **But be aware that generating Let's encrypt certificate on port different than 443 requires some additional server setup to satisfy the domain challenges** (check https://community.letsencrypt.org/t/create-or-generate-certificate-on-other-ports/124956).

In short, if you don't have other applications listening on 80/443, you can try running PocketBase like this:
```
./pocketbase serve --http="yourdomain.com:80" --https="yourdomain.com:443"
```
